### ISSUE TITLE
Failure previewing a subset of stack's tree.

### ISSUE THREAD

**Issue by timmyers** [pulumi/pulumi#4032]
Attempting to `preview` (and then `apply`) a subset of a stacks tree is resulting in an error.

For a stack that looks something like: 
```
❯ pulumi stack -u
...
    TYPE                                                                                NAME
    pulumi:pulumi:Stack                                                                 infrastructure-istio-operator
    │  URN: urn:pulumi:istio-operator::infrastructure::pulumi:pulumi:Stack::infrastructure-istio-operator
    ├─ infrastructure:k8s-istio-operator                                                cluster-us2
    │  │  URN: urn:pulumi:istio-operator::infrastructure::infrastructure:k8s-istio-operator::cluster-us2
    │  └─ kubernetes:yaml:ConfigGroup                                                   cluster-us2
    │     │  URN: urn:pulumi:istio-operator::infrastructure::infrastructure:k8s-istio-operator$kubernetes:yaml:ConfigGroup::cluster-us2
    │     └─ ...
    ├─ infrastructure:k8s-istio-operator                                                web-dev
    │  │  URN: urn:pulumi:istio-operator::infrastructure::infrastructure:k8s-istio-operator::web-dev
    │  └─ kubernetes:yaml:ConfigGroup                                                   web-dev
    │     │  URN: urn:pulumi:istio-operator::infrastructure::infrastructure:k8s-istio-operator$kubernetes:yaml:ConfigGroup::web-dev
    │     └─ ...
```

Trying `pulumi preview -t urn:pulumi:istio-operator::infrastructure::infrastructure:k8s-istio-operator::web-dev` results in this error
```
Diagnostics:
  infrastructure:k8s-istio-operator (web-dev):
    error: Cannot delete parent resource 'urn:pulumi:istio-operator::infrastructure::infrastructure:k8s-istio-operator::web-dev' without also deleting child 'urn:pulumi:istio-operator::infrastructure::infrastructure:k8s-istio-operator$kubernetes:yaml:ConfigGroup::web-dev'
```
Adding `--target-dependents` results in the same error.

```
❯ pulumi version
v1.12.0
```

**Comment by unthought** [comment]
same here, v1.14.0

**Comment by phewitt** [comment]
Seeing same behavior. Trying to delete a top level resource with many children using `--target` and `--target-dependents` and it warns of each child not included in `--target`

For example:

```
  eks:index:Cluster (myCluster):
    error: Cannot delete parent resource 'urn:pulumi:prod::infrastructure::eksCluster$eks:index:Cluster::myCluster' without also deleting child 'urn:pulumi:prod::infrastructure::eksCluster$eks:index:Cluster$eks:index:ServiceRole::myCluster-eksRole'.

```

**Comment by AgrimPrasad** [comment]
+1 Seeing this issue too, v2.2.1

**Comment by ppawiggers** [comment]
Still seeing this issue in v2.8.1

**Comment by alonikomax** [comment]
Exists in v3.1.0

**Comment by farvour** [comment]
Yeah this feature of pulumi has not, and still does not work correctly for me. Despite even having the right parent set for a sub-set/sub-stack of URNs. It's almost like pulumi's `--target-dependents` doesn't actually work when the resources have yet to be created in the stack so the whole operation just gives up thinking you're trying to delete the subs/children from the state.

**Comment by mitchellmaler** [comment]
Just ran into this same issue. We have multiple Components and would like to target at that level to target-replace or target. So far the only way I can make this work is to script out a way to grab all the components child resources to pass as a --target. Would be great is the --target-dependents would include all sub/child resources of Components.

**Comment by mitchellmaler** [comment]
@komalali I just tried this with Pulumi in python and it doesn't work with --target-replace. It did make it past the error noted above but all it did was replace the component resource and none of the children. Was this only added for target or should it also work for target-replace?

version: v3.10.1

**Comment by lukehoban** [comment]
@mitchellmaler It does sounds like there is an issue there that was not addressed with #7605.  Would you mind opening a new issue with details of the remaining problem related to `--target-replace`?
